bigdecimal/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+        ParseBigDecimalError::ParseDecimal(err)
+    }
+}
+
+impl From<ParseIntError> for ParseBigDecimalError {
+    fn from(_: ParseIntError) -> ParseBigDecimalError {
+        ParseBigDecimalError::Other
+    }
+}
+
+impl From<ParseBigIntError> for ParseBigDecimalError {
+    fn from(_: ParseBigIntError) -> ParseBigDecimalError {
+        // ParseBigDecimalError::ParseDecimal(err)
+        ParseBigDecimalError::Other
+    }
+}
+
+impl FromStr for BigDecimal {
+    type Err = ParseBigDecimalError;
+
+    #[inline]
+    fn from_str(s: &str) -> Result<BigDecimal, ParseBigDecimalError> {
+        BigDecimal::from_str_radix(s, 10)
+    }
+}
+
+impl PartialEq for BigDecimal {
+    #[inline]
+    fn eq(&self, rhs: &BigDecimal) -> bool {
+        // println!("{}E{} =?= {}E{}",
+        //          self.int_val,
+        //          self.scale,
+        //          rhs.int_val,
+        //          rhs.scale);
+
+        // difference in scale between the two decimals
+        let scale_diff = (self.scale - rhs.scale).abs() as u32;
+
+        // the scale as a bigint number power of ten
+        // e.g. diff == 3 -> scale_bigint == 1000
+        let scale_bigint = if scale_diff <= 9 {
+            BigInt::from_u64(10u64.pow(scale_diff)).unwrap()
+        } else {
+            let billion = &BigInt::from_u64(10u64.pow(9)).unwrap();
+
+            let mut scale_diff = scale_diff;
+            let mut tmp_bigint = BigInt::one();
+            while scale_diff > 9 {
+                tmp_bigint = tmp_bigint * billion;
+                scale_diff -= 9;
+            }
+
+            tmp_bigint * BigInt::from_u64(10u64.pow(scale_diff)).unwrap()
+        };
+
+        // fix scale and test equality
+        let result = if self.scale > rhs.scale {
+            let shifted_int = &rhs.int_val * scale_bigint;
+            shifted_int == self.int_val
+        } else if self.scale < rhs.scale {
+            let shifted_int = &self.int_val * scale_bigint;
+            shifted_int == rhs.int_val
+        } else {
+            self.int_val == rhs.int_val
+        };
+
+        return result;
+    }
+}
+
+impl Default for BigDecimal {
+    #[inline]
+    fn default() -> BigDecimal {
+        Zero::zero()
+    }
+}
+
+impl Zero for BigDecimal {
+    #[inline]
+    fn zero() -> BigDecimal {
+        BigDecimal::new(BigInt::zero(), 0)
+    }
+
+    #[inline]
+    fn is_zero(&self) -> bool {
+        self.int_val.is_zero()
+    }
+}
+
+impl One for BigDecimal {
+    #[inline]
+    fn one() -> BigDecimal {
+        BigDecimal::new(BigInt::one(), 0)
+    }
+}
+
+forward_all_binop_to_ref_ref!(impl Add for BigDecimal, add);
+
+impl<'a, 'b> Add<&'b BigDecimal> for &'a BigDecimal {
+    type Output = BigDecimal;
+
+    #[inline]
+    fn add(self, other: &BigDecimal) -> BigDecimal {
+        let scale = max(self.scale, other.scale);
+        let left = self.clone().set_scale(scale);
+        let right = other.clone().set_scale(scale);
```

Comment:
The math equations would benefit from a single clone (or rather "rescale") and then an inplace operator on the new object. It looks like `std::ops::AddAssign` was marked stable in 1.8; which means I cannot use that, right? Is there a way to write two implementations based on rustc version?